Fill out the how the web works learn content (#38109)
mdn/content: files/en-us/learn_web_development/getting_started/web_standards/how_browsers_load_websites/index.md

      <td>
        <ul>
          <li>The different kinds of assets that are returned in an HTTP response.</li>
          <li>How the different files are assembled by the browser to render a web page that is then displayed to the user.</li>
          <li>Why the browser is sometimes seen as a hostile programming environment, but also an awesome programming environment.</li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

## What files are returned in HTTP responses?

> To summarize the [overview of web technologies](/en-US/docs/Learn_web_development/Getting_started/Web_standards/The_web_standards_model#overview_of_modern_web_technologies) we looked at in the last article, HTTP responses (to requests for a web page) will generally contain some of the following file types:

Reviewer: Do you want to link "HTTP response" or is that too much information at this point?

Author: Yeah, TMI at this point.

- HTML files, which specify the web page content and its structure.
- CSS files, which specify styling and layout information.
- JavaScript files, which specify the behavior of interactive parts of the web page.
- Media assets such as images, videos, audio files, [PDFs](/en-US/docs/Glossary/PDF), and [SVGs](/en-US/docs/Glossary/SVG), which are embedded in web pages or otherwise displayed by the browser.
- Other kinds of file that the browser can't handle natively and therefore hands off to a relevant app on the device to render, for example Word or Pages documents, PowerPoint slide decks, and Open Office files.

> ## Web page rendereding

Reviewer: ```suggestion
## Web page rendering
```

> When the user navigates to a new web page (by clicking a link, or entering a web address in the browser address bar), several HTTP requests are sent, and several files are sent back in HTTP responses. The files received in these responses are processed by the browser and put together into a web page that the user can interact with. This prcess is called web page **rendering**.

Reviewer: ```suggestion
When the user navigates to a new web page (by clicking a link, or entering a web address in the browser address bar), several HTTP requests are sent, and several files are sent back in HTTP responses. The files received in these responses are processed by the browser and put together into a web page that the user can interact with. This process of assembling the pieces into a web page is called **rendering**.
```

## Handling JavaScript

Any JavaScript found on the page (either included in the HTML file, or fetched from external script files) is parsed, interpreted, compiled, and executed. This happens at some point before the final page rendering is completed — after all, some JavaScript may affect the rendering, for example by adding nodes to the DOM or modifying existing ones.

Returning to our example, let's say the following JavaScript is found in the HTML file:

```html hidden
<p>
  Let's use:
  <span>HTML</span>
  <span>CSS</span>
  <span>JavaScript</span>
</p>
```

```css hidden
span {
  border: 1px solid black;
  background-color: lime;
}
```

```js
const spans = document.querySelectorAll("span");
spans.forEach((span) => {
  const reversedText = span.textContent.split("").reverse().join("");
  span.textContent = reversedText;
});
```

You don't need to understand exactly how this JavaScript works, but at a high level, it finds every SPAN node in the DOM and reverses the order of the characters in their child text nodes.

The final output is as follows:

{{EmbedLiveSample('Handling the JavaScript', '100%', 90)}}

> ## What else happens?

Reviewer: ```suggestion
## What happens after HTML rendering?
```

Author: I'm not sure I agree with this suggestion. We'd at least need to say "What else happens after HTML rendering?", as we've already discussed other things that do, and also I don't think it really matters when they happen. It's just supposed to be a high-level idea.

Reviewer: Yes, gotcha (sorry I left out the rationale) I just find "What else happens?" a little vague as a H2, in general.

Author: That's fair. I've changed it to "What other rendering steps are there?"

> There are several other things that happen during page rendering, but we won't discuss them all here. One notable additional occurence worth mentioning is that an accessibility tree is built, based on the DOM, for assistive technologies (for example screen readers) to hook into, which enables people who are not able to see the rendered content to interact with it.

Reviewer: ```suggestion
There are several other things that happen during page rendering, but we won't discuss them all here. One notable additional occurrence worth mentioning is that an accessibility tree is built, based on the DOM, for assistive technologies (for example screen readers) to hook into, which enables people who are not able to see the rendered content to interact with it.
```

You will learn more about this later on, in our [Accessibility](/en-US/docs/Learn_web_development/Core/Accessibility) module.

## The browser: a hostile _and_ an awesome programming environment

Front-end web development can sometimes be frustrating, and some people consider the browser to be a hostile programming environment. This is because, unlike other programming environments, it is much harder to make guarantees about the environment your code will run on. You cannot know ahead of time all the different combinations of operating system, browser, language, location, network connection, CPU, GPU, memory, battery life, etc., that your users will have, therefore you cannot guarantee a perfect user experience for all of them.

> Modern browsers tend to implement web standards pretty consistently, but there is still a lot of uncertainty to navigate. As a web developer, you will need to embrace that uncertainty and learn to program defensively. This relies on adhering to the [best practices](/en-US/docs/Learn_web_development/Getting_started/Web_standards/The_web_standards_model#web_best_practices) outlined in the previous article.

Reviewer: ```suggestion
Modern browsers tend to implement web standards pretty consistently, but there is still a lot of uncertainty to navigate. As a web developer, you will need to embrace uncertainty and learn to program defensively and when to be conservative with the features you use. This relies on adhering to the [best practices](/en-US/docs/Learn_web_development/Getting_started/Web_standards/The_web_standards_model#web_best_practices) outlined in the previous article.
```

Author: I implemented this one, with a couple of tweaks.

On the upside, the web is also an awesome programming environment, for many reasons.

> - For a start, it is designed with univeral access in mind — its basic state is accessible and linkable. Some of these basics are harder to achieve in other environments.

Reviewer: ```suggestion
- For a start, it is designed with universal access in mind — its basic state is accessible and linkable. Some of these basics are harder to achieve in other environments.
```

- App delivery across the web is simple and powerful. You don't need to take your users through a complicated installation process: you just point them to a web address and off they go.

> - As a result, app updates are also easy — in many cases, you can just reload the browser tab. You don't need to worry about getting your users to constantly download and install large packages.

Reviewer: ```suggestion
- App updates are usually straightforward — in many cases, visitors can see new versions of an application when they reload their browser tab. You don't need to worry about figuring out how to get visitors regularly download and install large software updates.
```

Reviewer: perfect, tnx

> - The web community is vibrant and helpful. as we discuss later on in our [Research and learning](/en-US/docs/Learn_web_development/Getting_started/Soft_skills/Research_and_learning) article, there are lots of places you can do to ask for help, and great resources available to learn from.

Reviewer: ```suggestion
- The web community is vibrant and helpful. As we discuss later on in our [Research and learning](/en-US/docs/Learn_web_development/Getting_started/Soft_skills/Research_and_learning) article, there are lots of places you can do to ask for help, and great resources available to learn from.
```
